File: LitmusAgent/src/browser/HtmlCleaner.ts
import { Page } from 'playwright';
import { logger } from '../utils/logger';
import { DEFAULT_WORD_COUNT_LIMIT } from '../config/constants';

/**
 * Configuration for HTML cleaning
 */
export interface HtmlCleanerConfig {
    removeHead: boolean;
    removeScripts: boolean;
    removeStyleAttributes: boolean;
    removeInlineStyles: boolean;
    removeComments: boolean;
    removeEmptyElements: boolean;
    removeEventHandlers: boolean;
    preserveTextContent: boolean;
}

/**
 * Default configuration for HTML cleaning
 */
export const DEFAULT_HTML_CLEANER_CONFIG: HtmlCleanerConfig = {
    removeHead: true,
    removeScripts: true,
    removeStyleAttributes: true,
    removeInlineStyles: true,
    removeComments: true,
    removeEmptyElements: false,
    removeEventHandlers: true,
    preserveTextContent: true
};

/**
 * HTML Cleaner utility for cleaning HTML content before AI processing
 */
export class HtmlCleaner {
    private config: HtmlCleanerConfig;

    constructor(config: Partial<HtmlCleanerConfig> = {}) {
        this.config = { ...DEFAULT_HTML_CLEANER_CONFIG, ...config };
    }

    /**
     * Clean HTML content by removing specified elements and attributes
     * @param html The HTML content to clean
     * @returns Cleaned HTML content
     */
    public cleanHtml(html: string): string {
        try {
            let cleanedHtml = html;

            // Remove head section if configured
            if (this.config.removeHead) {
                cleanedHtml = this.removeHeadSection(cleanedHtml);
            }

            // Remove script tags if configured
            if (this.config.removeScripts) {
                cleanedHtml = this.removeScriptTags(cleanedHtml);
            }

            // Remove style attributes if configured
            if (this.config.removeStyleAttributes) {
                cleanedHtml = this.removeStyleAttributes(cleanedHtml);
            }

            // Remove inline styles if configured
            if (this.config.removeInlineStyles) {
                cleanedHtml = this.removeInlineStyles(cleanedHtml);
            }

            // Remove comments if configured
            if (this.config.removeComments) {
                cleanedHtml = this.removeComments(cleanedHtml);
            }

            // Remove empty elements if configured
            if (this.config.removeEmptyElements) {
                cleanedHtml = this.removeEmptyElements(cleanedHtml);
            }

            // Remove event handlers if configured
            if (this.config.removeEventHandlers) {
                cleanedHtml = this.removeEventHandlers(cleanedHtml);
            }

            logger.info('HTML cleaned successfully');
            return cleanedHtml;
        } catch (error) {
            logger.error('Error cleaning HTML:', error);
            throw error;
        }
    }

    /**
     * Remove the entire head section from HTML
     */
    private removeHeadSection(html: string): string {
        return html.replace(/<head[^>]*>[\s\S]*?<\/head>/gi, '');
    }

    /**
     * Remove all script tags and their content
     */
    private removeScriptTags(html: string): string {
        return html.replace(/<script[^>]*>[\s\S]*?<\/script>/gi, '');
    }

    /**
     * Remove style attributes from all elements
     */
    private removeStyleAttributes(html: string): string {
        return html.replace(/\sstyle\s*=\s*["'][^"']*["']/gi, '');
    }

    /**
     * Remove inline style tags and their content
     */
    private removeInlineStyles(html: string): string {
        return html.replace(/<style[^>]*>[\s\S]*?<\/style>/gi, '');
    }

    /**
     * Remove HTML comments
     */
    private removeComments(html: string): string {
        return html.replace(/<!--[\s\S]*?-->/g, '');
    }

    /**
     * Remove empty elements (elements with no content and no attributes)
     */
    private removeEmptyElements(html: string): string {
        // Remove empty divs, spans, p tags, etc.
        return html.replace(/<(div|span|p|section|article|header|footer|nav|aside|main)[^>]*>\s*<\/\1>/gi, '');
    }

    /**
     * Remove event handler attributes (onclick, onload, etc.)
     */
    private removeEventHandlers(html: string): string {
        // Remove common event handler attributes - handle double and single quotes separately
        let cleanedHtml = html;
        // Remove double-quoted event handlers
        cleanedHtml = cleanedHtml.replace(/\s*(onclick|onload|onchange|onsubmit|onmouseover|onmouseout|onfocus|onblur|onkeydown|onkeyup|onkeypress)\s*=\s*"[^"]*"/gi, '');
        // Remove single-quoted event handlers
        cleanedHtml = cleanedHtml.replace(/\s*(onclick|onload|onchange|onsubmit|onmouseover|onmouseout|onfocus|onblur|onkeydown|onkeyup|onkeypress)\s*=\s*'[^']*'/gi, '');
        return cleanedHtml;
    }

    /**
     * Get page HTML and clean it using page.evaluate
     * @param page The Playwright page object
     * @param wordCountLimit Maximum number of words to include in the cleaned HTML (default: 10000)
     * @returns Cleaned HTML content
     */
    public async getCleanedPageHtml(page: Page, wordCountLimit: number = DEFAULT_WORD_COUNT_LIMIT): Promise<string> {
        try {
            const cleanedHtml = await page.evaluate(({ config, wordLimit }) => {
                // Get the full HTML content
                let html = document.documentElement.outerHTML;

                // Remove head section if configured
                if (config.removeHead) {
                    html = html.replace(/<head[^>]*>[\s\S]*?<\/head>/gi, '');
                }

                // Remove script tags if configured
                if (config.removeScripts) {
                    html = html.replace(/<script[^>]*>[\s\S]*?<\/script>/gi, '');
                }

                // Remove style attributes if configured
                if (config.removeStyleAttributes) {
                    html = html.replace(/\sstyle\s*=\s*["'][^"']*["']/gi, '');
                }

                // Remove inline style tags if configured
                if (config.removeInlineStyles) {
                    html = html.replace(/<style[^>]*>[\s\S]*?<\/style>/gi, '');
                }

                // Remove comments if configured
                if (config.removeComments) {
                    html = html.replace(/<!--[\s\S]*?-->/g, '');
                }

                // Remove empty elements if configured
                if (config.removeEmptyElements) {
                    html = html.replace(/<(div|span|p|section|article|header|footer|nav|aside|main)[^>]*>\s*<\/\1>/gi, '');
                }

                // Remove event handlers if configured
                if (config.removeEventHandlers) {
                    // Remove double-quoted event handlers
                    html = html.replace(/\s*(onclick|onload|onchange|onsubmit|onmouseover|onmouseout|onfocus|onblur|onkeydown|onkeyup|onkeypress)\s*=\s*"[^"]*"/gi, '');
                    // Remove single-quoted event handlers
                    html = html.replace(/\s*(onclick|onload|onchange|onsubmit|onmouseover|onmouseout|onfocus|onblur|onkeydown|onkeyup|onkeypress)\s*=\s*'[^']*'/gi, '');
                }

                // Apply word count limit if specified
                if (wordLimit && wordLimit > 0) {
                    // Extract text content and count words
                    const tempDiv = document.createElement('div');
                    tempDiv.innerHTML = html;
                    const textContent = tempDiv.textContent || tempDiv.innerText || '';
                    const words = textContent.trim().split(/\s+/).filter(word => word.length > 0);
                    
                    if (words.length > wordLimit) {
                        // If word count exceeds limit, truncate the HTML
                        // We'll truncate by removing content from the end while preserving structure
                        const truncatedWords = words.slice(0, wordLimit);
                        const truncatedText = truncatedWords.join(' ');
                        
                        // Simple approach: find the position in the original HTML where we need to cut
                        let currentWordCount = 0;
                        let cutPosition = 0;
                        
                        // Walk through the HTML character by character to find the cut point
                        const htmlText = tempDiv.textContent || '';
                        const htmlWords = htmlText.trim().split(/\s+/).filter(word => word.length > 0);
                        
                        if (htmlWords.length > wordLimit) {
                            // Find the position of the last word we want to keep
                            const wordsToKeep = htmlWords.slice(0, wordLimit);
                            const textToKeep = wordsToKeep.join(' ');
                            
                            // Find this text in the original HTML
                            const textIndex = html.indexOf(textToKeep);
                            if (textIndex !== -1) {
                                cutPosition = textIndex + textToKeep.length;
                            } else {
                                // Fallback: estimate position based on word ratio
                                const wordRatio = wordLimit / htmlWords.length;
                                cutPosition = Math.floor(html.length * wordRatio);
                            }
                            
                            // Truncate the HTML at the calculated position
                            html = html.substring(0, cutPosition);
                            
                            // Ensure we don't cut in the middle of a tag
                            const lastOpenTag = html.lastIndexOf('<');
                            const lastCloseTag = html.lastIndexOf('>');
                            
                            if (lastOpenTag > lastCloseTag) {
                                // We cut in the middle of a tag, remove the incomplete tag
                                html = html.substring(0, lastOpenTag);
                            }
                        }
                    }
                }

                return html;
            }, { config: this.config, wordLimit: wordCountLimit });

            logger.info('Page HTML cleaned successfully');
            return cleanedHtml;
        } catch (error) {
            logger.error('Error getting cleaned page HTML:', error);
            throw error;
        }
    }

    /**
     * Update the cleaner configuration
     * @param newConfig Partial configuration to update
     */
    public updateConfig(newConfig: Partial<HtmlCleanerConfig>): void {
        this.config = { ...this.config, ...newConfig };
        logger.info('HTML cleaner configuration updated');
    }

    /**
     * Get the current configuration
     * @returns Current configuration
     */
    public getConfig(): HtmlCleanerConfig {
        return { ...this.config };
    }
}

File: src/utils/html_cleaner.py
"""
HTML Cleaner utility for cleaning HTML content before AI processing.
Port of LitmusAgent's HtmlCleaner.ts to Python.
"""

import re
from typing import Optional
from log_config.logger import logger

DEFAULT_WORD_COUNT_LIMIT = 10000


def clean_html(
    html: str,
    remove_head: bool = True,
    remove_scripts: bool = True,
    remove_style_attributes: bool = True,
    remove_inline_styles: bool = True,
    remove_comments: bool = True,
    remove_event_handlers: bool = True,
    word_count_limit: Optional[int] = DEFAULT_WORD_COUNT_LIMIT
) -> str:
    """
    Clean HTML content by removing specified elements and attributes.
    
    Args:
        html: The HTML content to clean
        remove_head: Remove <head> section
        remove_scripts: Remove <script> tags
        remove_style_attributes: Remove style="..." attributes
        remove_inline_styles: Remove <style> tags
        remove_comments: Remove HTML comments
        remove_event_handlers: Remove onclick, onload, etc.
        word_count_limit: Maximum words to keep (None for no limit)
    
    Returns:
        Cleaned HTML content
    """
    try:
        cleaned_html = html
        
        # Remove head section
        if remove_head:
            cleaned_html = re.sub(r'<head[^>]*>[\s\S]*?</head>', '', cleaned_html, flags=re.IGNORECASE)
        
        # Remove script tags
        if remove_scripts:
            cleaned_html = re.sub(r'<script[^>]*>[\s\S]*?</script>', '', cleaned_html, flags=re.IGNORECASE)
        
        # Remove style attributes
        if remove_style_attributes:
            cleaned_html = re.sub(r'\sstyle\s*=\s*["\'][^"\']*["\']', '', cleaned_html, flags=re.IGNORECASE)
        
        # Remove inline style tags
        if remove_inline_styles:
            cleaned_html = re.sub(r'<style[^>]*>[\s\S]*?</style>', '', cleaned_html, flags=re.IGNORECASE)
        
        # Remove HTML comments
        if remove_comments:
            cleaned_html = re.sub(r'<!--[\s\S]*?-->', '', cleaned_html)
        
        # Remove event handlers
        if remove_event_handlers:
            event_handlers = [
                'onclick', 'onload', 'onchange', 'onsubmit', 
                'onmouseover', 'onmouseout', 'onfocus', 'onblur',
                'onkeydown', 'onkeyup', 'onkeypress'
            ]
            pattern = r'\s*(' + '|'.join(event_handlers) + r')\s*=\s*["\'][^"\']*["\']'
            cleaned_html = re.sub(pattern, '', cleaned_html, flags=re.IGNORECASE)
        
        # Apply word count limit
        if word_count_limit and word_count_limit > 0:
            cleaned_html = _apply_word_limit(cleaned_html, word_count_limit)
        
        logger.info(f'HTML cleaned successfully. Length: {len(cleaned_html)} chars')
        return cleaned_html
        
    except Exception as e:
        logger.error(f'Error cleaning HTML: {str(e)}')
        raise


def _apply_word_limit(html: str, word_limit: int) -> str:
    """Truncate HTML to approximately word_limit words."""
    # Extract text and count words
    text_only = re.sub(r'<[^>]+>', ' ', html)
    words = text_only.split()
    
    if len(words) <= word_limit:
        return html
    
    # Estimate cut position based on word ratio
    word_ratio = word_limit / len(words)
    cut_position = int(len(html) * word_ratio)
    
    truncated = html[:cut_position]
    
    # Ensure we don't cut in the middle of a tag
    last_open = truncated.rfind('<')
    last_close = truncated.rfind('>')
    
    if last_open > last_close:
        truncated = truncated[:last_open]
    
    logger.info(f'HTML truncated from {len(words)} to ~{word_limit} words')
    return truncated

